pkg/controller/node/node_controller.go

```diff
@@ -0,0 +1,131 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	cloudprovider "github.com/openshift/cloud-network-config-controller/pkg/cloudprovider"
+	controller "github.com/openshift/cloud-network-config-controller/pkg/controller"
+)
+
+var (
+	// nodeControllerAgentType is the Node controller's dedicated resource type
+	nodeControllerAgentType reflect.Type = reflect.TypeOf(&corev1.Node{})
+	// nodeControllerAgentName is the controller name for the Node controller
+	nodeControllerAgentName = "node"
+	// nodeEgressIPConfigAnnotationKey is the annotation key used for indicating the node's egress IP configuration
+	nodeEgressIPConfigAnnotationKey = "cloud.network.openshift.io/egress-ipconfig"
+)
+
+// NodeController is the controller implementation for Node resources
+// This controller is used to annotate nodes for the purposes of the
+// cloud network config controller
+type NodeController struct {
+	// Implements its own Node lister
+	nodesLister corelisters.NodeLister
+	// CloudProviderClient is a client interface allowing the controller
+	// access to the cloud API
+	CloudProviderClient cloudprovider.CloudProviderIntf
+	// KubeClientset is a standard kubernetes clientset
+	KubeClientset kubernetes.Interface
+}
+
+// NewNodeController returns a new Node controller
+func NewNodeController(
+	kubeClientset kubernetes.Interface,
+	cloudProviderClient cloudprovider.CloudProviderIntf,
+	nodeInformer coreinformers.NodeInformer) *controller.CloudNetworkConfigController {
+
+	nodeController := &NodeController{
+		nodesLister:         nodeInformer.Lister(),
+		KubeClientset:       kubeClientset,
+		CloudProviderClient: cloudProviderClient,
+	}
+
+	controller := controller.NewCloudNetworkConfigController(
+		[]cache.InformerSynced{nodeInformer.Informer().HasSynced},
+		nodeController,
+		nodeControllerAgentName,
+		nodeControllerAgentType,
+	)
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.Enqueue,
+	})
+	return controller
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Node resource
+// with the current status of the resource.
+func (n *NodeController) SyncHandler(key string) error {
+	// Convert the key to a name (since Node is non-namespaced)
+	_, name, err := cache.SplitMetaNamespaceKey(key)
```

Comment:
for non-namespaced objects, `cache.MetaNamespaceKeyFunc()` just returns the name anyway... there's no splitting to be done here. (likewise in some other places)